libcontainer/rootfs_linux.go

```diff
@@ -1115,3 +1120,51 @@ func setRecAttr(m *configs.Mount, rootfs string) error {
 		return unix.MountSetattr(-1, procfd, unix.AT_RECURSIVE, m.RecAttr)
 	})
 }
+
+// validateRoot detects symlinks in config.Rootfs that may break config.ReadonlyPaths
+// and config.MaskPaths.
+//
+// For preventing CVE-2023-27561 (CVE-2019-19921 re-introduction/regression).
+// https://github.com/opencontainers/runc/issues/3751
+func validateRoot(config *configs.Config) error {
+	mustNotBeSymlink := make(map[string]struct{})
+	for _, f := range append(config.ReadonlyPaths, config.MaskPaths...) {
+		mustNotBeSymlink[f] = struct{}{}
+		a, err := ancestorPaths(f)
+		if err != nil {
+			return fmt.Errorf("failed to compute the ancestor paths of %q: %w", f, err)
+		}
+		for _, g := range a {
+			mustNotBeSymlink[g] = struct{}{}
+		}
+	}
+	// mustNotBeSymlink typically contains {"/sys", "/proc"} in addition to readonlyPaths and maskedPaths.
+	for f := range mustNotBeSymlink {
+		if f != filepath.Clean(f) {
```

Comment:
It seems that  this check was already performed in `ancestorPaths` for all the paths. What's the point of repeating it?

Reply:
`mustNotBeSymlink[f] = struct{}{}` in L1132 is not covered by that check

Reply:
In fact it is covered, since `f` is passed on to `ancestorPaths` and the first thing it is doing is checking if the path is clean